pkg/cloud/azure/azure.go

```diff
@@ -0,0 +1,211 @@
+	nattPort, nattDiscoveryPort, gateways int,
+) (*azureProvider, error) {
+	if infraID == "" {
+		return nil, fmt.Errorf("cluster infraID is empty")
+	}
+
+	if nattPort == 0 {
+		nattPort = constants.SubmarinerNatTPort
+	}
+
+	if instanceType == "" {
+		instanceType = gwInstanceType
+	}
+
+	if gateways < 1 {
+		return nil, fmt.Errorf("the count of gateways is less than 1")
+	}
+
+	if nattDiscoveryPort == 0 {
+		nattDiscoveryPort = constants.SubmarinerNatTDiscoveryPort
+	}
+
+	subscriptionID, err := initializeFromAuthFile(hubKubeClient, clusterName, credentialsSecretName)
+
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+
+	if err != nil {
+		klog.Errorf("unable to create the Azure  authorizer :%v", err)
+		return nil, err
+	}
+
+	k8sClient := k8s.NewInterface(kubeClient)
+
+	msDeployer := ocp.NewK8sMachinesetDeployer(restMapper, dynamicClient)
+
+	cloudInfo := azure.CloudInfo{
+		SubscriptionID: subscriptionID,
+		InfraID:        infraID,
+		Region:         region,
+		BaseGroupName:  infraID + "-rg",
+		Authorizer:     authorizer,
+		K8sClient:      k8sClient,
+	}
+
+	azureCloud := azure.NewCloud(&cloudInfo)
+
+	gwDeployer, err := azure.NewOcpGatewayDeployer(&cloudInfo, azureCloud, msDeployer, instanceType, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to initialize a GatewayDeployer config")
+	}
+
+	cloudPrepare := azure.NewCloud(&cloudInfo)
+
+	return &azureProvider{
+		infraID:              infraID,
+		nattPort:             uint16(nattPort),
+		routePort:            strconv.Itoa(constants.SubmarinerRoutePort),
+		gatewayMetricsPort:   constants.SubmarinerGatewayMetricsPort,
+		globalnetMetricsPort: constants.SubmarinerGlobalnetMetricsPort,
+		cloudPrepare:         cloudPrepare,
+		gwDeployer:           gwDeployer,
+		reporter:             reporter.NewEventRecorderWrapper("AzureCloudProvider", eventRecorder),
+		nattDiscoveryPort:    int64(nattDiscoveryPort),
+		gateways:             gateways,
+	}, nil
+}
+
+// PrepareSubmarinerClusterEnv prepares submariner cluster environment on Azure
+// The below tasks will be executed
+// 1. create the inbound and outbound firewall rules for submariner, below ports will be opened
+//    - NAT traversal port (by default 4500/UDP)
+//    - 4800/UDP port to encapsulate Pod traffic from worker and master nodes to the Submariner Gateway nodes
+//    - ESP & AH protocols for private-ip to private-ip gateway communications
+// 2. create the inbound and outbound firewall rules to open 8080/TCP, 8081/TCP port to export metrics service from the
+//    Submariner gateway
+func (r *azureProvider) PrepareSubmarinerClusterEnv() error {
+	if err := r.gwDeployer.Deploy(api.GatewayDeployInput{
+		PublicPorts: []api.PortSpec{
+			{Port: r.nattPort, Protocol: "udp"},
+			{Port: uint16(r.nattDiscoveryPort), Protocol: "udp"},
+			{Port: 0, Protocol: "esp"},
+			{Port: 0, Protocol: "ah"},
+		},
+		Gateways: r.gateways,
+	}, r.reporter); err != nil {
+		return err
+	}
+
```

Comment:
Please add a "TODO:" note here saying that for OVN Clusters, we have to avoid opening the UDP/4800 ports. @vthapar when working on the necessary OVN PR can modify this part as well.

Reply:
Done